src/main/java/net/shadowfacts/shadowlib/version/Version.java

```diff
@@ -0,0 +1,367 @@
+		}
+
+		major = Integer.parseInt(arr2[0]);
+		minor = Integer.parseInt(arr2[1]);
+		patch = Integer.parseInt(arr2[2]);
+
+
+		if (arr.length == 2) { // There is 1 label
+
+			if (arr[1] == null) {
+				throw new InvalidVersionException("Cannot create Version with null label");
+			} else if (arr[1].isEmpty()) {
+				throw new InvalidVersionException("Cannot create Version with empty label");
+			}
+
+			label = arr[1];
+
+		} else if (arr.length < 2) { //  There are multiple labels
+
+			for (int i = 1; i < arr.length - 1; i++) {
+
+				if (arr[i] == null) {
+					throw new InvalidVersionException("Cannot create Version with null label");
+				} else if (arr[i].equals("")) {
+					throw new InvalidVersionException("Cannot create Version with empty label");
+				}
+
+				label += arr[i];
+
+			}
+
+		}
+	}
+
+	/**
+	 * Creates a new Version object without a label from major, minor, and patch numbers
+	 * @param major
+	 * @param minor
+	 * @param patch
+	 */
+	public Version(int major, int minor, int patch) {
+		this.major = major;
+		this.minor = minor;
+		this.patch = patch;
+	}
+
+	/**
+	 * Creates a new {@link Version} object from a major version, minor version, patch number, and label
+	 * @param major
+	 * @param minor
+	 * @param patch
+	 * @param label
+	 */
+	public Version(int major, int minor, int patch, String label) {
+		this.major = major;
+		this.minor = minor;
+		this.patch = patch;
+		this.label = label;
+	}
+
+	/**
+	 * Creates a new {@link Version} from the current {@link Version} object
+	 * @return The duplicated {@link Version}
+	 */
+	public Version copy() {
+		return new Version(major, minor, patch, label);
+	}
+
+	/**
+	 * Increments the MAJOR version by 1
+	 * MINOR and PATCH are reset to 0
+	 * @param resetLabel If true, removes the label
+	 * @return
+	 */
+	public Version incrementMajor(boolean resetLabel) {
+		major++;
+		minor = 0;
+		patch = 0;
+		if (resetLabel) {
+			label = "";
+		}
+		return this;
+	}
+
+	/**
+	 * Increments the MAJOR version by 1
+	 * MINOR and PATCH are reset to 0
+	 * Does not remove the label
+	 * @return
+	 */
+	public Version incrementMajor() {
+		return incrementMajor(false);
+	}
+
+	/**
+	 * Increments the MINOR version by 1
+	 * PATCH is reset to 0
+	 * @param resetLabel If true, removes the label
+	 * @return
+	 */
+	public Version incrementMinor(boolean resetLabel) {
+		minor++;
+		patch = 0;
+		if (resetLabel) {
+			label = "";
+		}
+		return this;
+	}
+
+	/**
+	 * Increments the MINOR version by 1
+	 * PATCH is reset to 0
+	 * Does not remote the label
+	 * @return
+	 */
+	public Version incrementMinor() {
+		return incrementMinor(false);
+	}
+
+	/**
+	 * Increments the PATCH version by 1
+	 * @param resetLabel If true, removes the label
+	 * @return
+	 */
+	public Version incrementPatch(boolean resetLabel) {
+		patch++;
+		if (resetLabel) {
+			label = "";
+		}
+		return this;
+	}
```

Comment:
They're not, they're just part of the `Version` class in ShadowLib. If we end up going with weak SemVer instead of strict, this'll get replaced by something else.